xtask/src/main.rs

```diff
@@ -0,0 +1,221 @@
+use std::{
+    cmp::Ordering,
+    collections::HashMap,
+    env, fs,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+type DynError = Box<dyn std::error::Error>;
+
+fn main() {
+    if let Err(e) = try_main() {
+        eprintln!("{}", e);
+        std::process::exit(-1);
+    }
+}
+
+fn try_main() -> Result<(), DynError> {
+    let task = env::args().nth(1);
+    match task.as_deref() {
+        Some("build") => build()?,
+        _ => print_help(),
+    }
+    Ok(())
+}
+
+fn print_help() {
+    eprintln!(
+        "Tasks:
+
+wasm            build wasm using wasm32-wasi target
+components      build components using wasm-tools
+"
+    )
+}
+
+#[derive(Debug, Clone)]
+struct Guest {
+    path: PathBuf,
+    name: String,
+    name_output: String,
+}
+
+impl Guest {
+    fn new(path: PathBuf, name: &str, name_output: Option<&str>) -> Self {
+        let name_output = name_output.unwrap_or(name).to_string();
+        let name = name.to_string();
+        return Self {
+            path,
+            name,
+            name_output,
+        };
+    }
+}
+
+// 1. Read all contents of the ./guests/ directory
+// 2. Filter out all non-directories (like README.md)
+// 3. Map the remaining paths to Guest structs containing its:
+//   - path: used for build
+//   - name: used for Error messages
+//   - name_output: used for build when defined in `name_map`
+// 4. Order the items by priority. Because packages like `function` should be built last
+fn guests() -> Vec<Guest> {
+    let dir = fs::read_dir(&dir_guests()).unwrap();
+    let name_map = HashMap::from([("function", "mycelia_guest_function")]);
+    let priority = vec!["*".to_string(), "function".to_string()];
+
+    let mut guests_filtered = dir
+        .map(|p| p.unwrap().path())
+        .filter(|p| p.is_dir())
+        .map(|p| {
+            let name = p.strip_prefix(&dir_guests()).unwrap().to_str().unwrap();
+            let name_output = name_map.get(name);
+            return Guest::new(p.clone(), name, name_output.copied());
+        })
+        .collect::<Vec<_>>();
+
+    guests_filtered.sort_by(|a, b| {
+        let a = priority.iter().position(|p| p == &a.name).unwrap_or(0);
+        let b = priority.iter().position(|p| p == &b.name).unwrap_or(0);
+        if a > b {
+            return Ordering::Greater;
+        } else if a < b {
+            return Ordering::Less;
+        } else {
+            return Ordering::Equal;
+        }
+    });
+
+    return guests_filtered;
+}
+
+fn build() -> Result<(), DynError> {
+    let _ = fs::remove_dir_all(&dir_target());
```

Comment:
Ahh I think I see whats going on with your comment regarding building times now.

There shouldn't be a need to nuke the entire `targets` directory. Cargo is good about maintaining it's cache.